File: hw2/hw2.py
import math
import random
def read_input_file(filename):
    with open(filename, 'r') as file:
        lines = file.readlines()
        n = int(lines[0])
        tau = float(lines[1])
        messages = []
        for line in lines[2:]:
            parts = line.strip().split()
            priority = int(parts[0])
            transmission_time = float(parts[1])
            period = int(parts[2])
            messages.append((priority, transmission_time, period))
        return n, tau, messages

def findworstcase(n, tau, messages):
    print("Number of messages:", n)
    print("Tau:", tau)
    min_sum = 1000
    min_message = None
    num_iterations = 1280000
    for _ in range(num_iterations):  # Replace 5 with the number of times you want to execute the loop
        priorities = [priority for priority, _, _ in messages]
        rest_of_messages = [(transmission_time, period) for _, transmission_time, period in messages]
        # Shuffle the priorities
        random.shuffle(priorities)
        # Reconstruct the messages with shuffled priorities
        shuffled_messages = [(priority,) + rest for priority, rest in zip(priorities, rest_of_messages)]
        #sorted_messages = sorted(shuffled_messages, key=lambda message: message[0])
        #print(shuffled_messages)
        #print("shuffled_messages:")
        final_sum = 0
        for i, (priority, transmission_time, period) in enumerate(shuffled_messages):
            #print(f"Message {i+1}: Priority={priority}, Transmission Time={transmission_time}, Period={period}")
            max = 0
            for j, (priority_1, transmission_time_1, period_1) in enumerate(shuffled_messages):
                if priority_1 >= priority:
                    if transmission_time_1 > max:
                        max = transmission_time_1
            #print(max)
            sum = 0
            q = max
            for j, (priority_1, transmission_time_1, period_1) in enumerate(shuffled_messages):
                if priority_1 < priority:
                    sum += math.ceil((q + tau) / period_1) * transmission_time_1
            while(q != max + sum and (max + sum + transmission_time) <= period):
                q = max + sum
                sum = 0
                for j, (priority_1, transmission_time_1, period_1) in enumerate(shuffled_messages):
                    if priority_1 < priority:
                        #print((q + tau) / period_1)
                        sum += math.ceil((q + tau) / period_1) * transmission_time_1
            if (max + sum + transmission_time) > period:
                #print("invalid")
                final_sum = -1
                break
            else:
                #print(f"{(q + transmission_time):.2f}")
                final_sum += q + transmission_time
        #print(f"Final sum: {final_sum:.2f}")
        if final_sum != -1 and final_sum < min_sum:
            min_sum = final_sum
            min_message = [pr[0] for pr in shuffled_messages]
    print(f"Minimum sum: {min_sum:.2f}")
    print(f"Priority with minimum sum: {min_message}")
            
input_filename = "input.dat"
n, tau, messages = read_input_file(input_filename)
findworstcase(n, tau, messages)

File: hw2/hw2.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

typedef struct {
    int priority;
    double transmission_time;
    int period;
} Message;

int read_input_file(const char *filename, int *n, double *tau, Message **messages) {
    FILE *file = fopen(filename, "r");
    if (!file) {
        perror("Error opening file");
        return 1;
    }

    fscanf(file, "%d", n);
    fscanf(file, "%lf", tau);

    *messages = (Message *)malloc(*n * sizeof(Message));
    if (!*messages) {
        perror("Memory allocation failed");
        fclose(file);
        return 1;
    }

    for (int i = 0; i < *n; ++i) {
        fscanf(file, "%d %lf %d", &((*messages)[i].priority), &((*messages)[i].transmission_time), &((*messages)[i].period));
    }

    fclose(file);
    return 0;
}

void findworstcase(int n, double tau, Message *messages) {
    printf("Number of messages: %d\n", n);
    printf("Tau: %f\n", tau);
    double min_sum = 1000;
    int *min_message = (int *)malloc(n * sizeof(int));
    int num_iterations = 2000000;
    srand(time(NULL));

    for (int iter = 0; iter < num_iterations; ++iter) {
        // Shuffle priorities
        for (int i = 0; i < n; ++i) {
            int j = rand() % n;
            int temp = messages[i].priority;
            messages[i].priority = messages[j].priority;
            messages[j].priority = temp;
        }

        double final_sum = 0;
        for (int i = 0; i < n; ++i) {
            double max = 0;
            for (int j = 0; j < n; ++j) {
                if (messages[j].priority >= messages[i].priority) {
                    if (messages[j].transmission_time > max) {
                        max = messages[j].transmission_time;
                    }
                }
            }

            double sum = 0;
            double q = max;
            for (int j = 0; j < n; ++j) {
                if (messages[j].priority < messages[i].priority) {
                    sum += ceil((q + tau) / messages[j].period) * messages[j].transmission_time;
                }
            }

            while (q != max + sum && (max + sum + messages[i].transmission_time) <= messages[i].period) {
                q = max + sum;
                sum = 0;
                for (int j = 0; j < n; ++j) {
                    if (messages[j].priority < messages[i].priority) {
                        sum += ceil((q + tau) / messages[j].period) * messages[j].transmission_time;
                    }
                }
            }

            if ((max + sum + messages[i].transmission_time) > messages[i].period) {
                final_sum = -1;
                break;
            } else {
                final_sum += q + messages[i].transmission_time;
            }
        }

        if (final_sum != -1 && final_sum < min_sum) {
            min_sum = final_sum;
            for (int i = 0; i < n; ++i) {
                min_message[i] = messages[i].priority;
            }
        }
    }

    printf("Minimum sum: %f\n", min_sum);
    printf("Priority with minimum sum: ");
    for (int i = 0; i < n; ++i) {
        printf("%d ", min_message[i]);
    }
    printf("\n");

    free(min_message);
}

int main() {
    const char *input_filename = "input.dat";
    int n;
    double tau;
    Message *messages;

    if (read_input_file(input_filename, &n, &tau, &messages) == 0) {
        findworstcase(n, tau, messages);
        free(messages);
    }

    return 0;
}